File: backend/src/main/java/com/jobportal/service/UserService.java
package com.jobportal.service;

import com.jobportal.dto.UpdateProfileRequest;
import com.jobportal.dto.UpdateUserSettingsRequest;
import com.jobportal.dto.UserProfileResponse;
import com.jobportal.dto.UserSettingsResponse;
import com.jobportal.entity.User;
import com.jobportal.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
public class UserService {

    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;

    public User getUserByEmail(String email) {
        return userRepository.findByEmail(email)
                .orElseThrow(() -> new RuntimeException("User not found"));
    }

    public UserProfileResponse getProfile(String email) {
        return mapToProfile(getUserByEmail(email));
    }

    @Transactional
    public UserProfileResponse updateProfile(String email, UpdateProfileRequest request) {
        User user = getUserByEmail(email);

        if (request.getFullName() != null && !request.getFullName().isBlank()) {
            user.setName(request.getFullName().trim());
        }
        if (request.getPhone() != null) {
            user.setPhone(request.getPhone().trim());
        }
        if (request.getLocation() != null) {
            user.setLocation(request.getLocation().trim());
        }
        if (request.getBio() != null) {
            user.setBio(request.getBio().trim());
        }
        if (request.getSkills() != null) {
            user.setSkills(request.getSkills().trim());
        }
        if (request.getExperience() != null) {
            user.setExperience(request.getExperience().trim());
        }
        if (request.getEducation() != null) {
            user.setEducation(request.getEducation().trim());
        }

        return mapToProfile(userRepository.save(user));
    }

    public UserSettingsResponse getSettings(String email) {
        return mapToSettings(getUserByEmail(email));
    }

    @Transactional
    public UserSettingsResponse updateSettings(String email, UpdateUserSettingsRequest request) {
        User user = getUserByEmail(email);

        if (request.getEmail() != null && !request.getEmail().isBlank()) {
            String newEmail = request.getEmail().trim();
            if (!newEmail.equalsIgnoreCase(user.getEmail())) {
                if (userRepository.existsByEmail(newEmail)) {
                    throw new RuntimeException("Email is already in use");
                }
                user.setEmail(newEmail);
            }
        }

        boolean changingPassword = request.getNewPassword() != null && !request.getNewPassword().isBlank();
        if (changingPassword) {
            if (request.getCurrentPassword() == null || request.getCurrentPassword().isBlank()) {
                throw new RuntimeException("Current password is required to set a new password");
            }
            if (!passwordEncoder.matches(request.getCurrentPassword(), user.getPassword())) {
                throw new RuntimeException("Current password is incorrect");
            }
            if (request.getConfirmPassword() == null
                    || !request.getNewPassword().equals(request.getConfirmPassword())) {
                throw new RuntimeException("New password and confirmation do not match");
            }
            if (request.getNewPassword().length() < 6) {
                throw new RuntimeException("New password must be at least 6 characters");
            }
            user.setPassword(passwordEncoder.encode(request.getNewPassword()));
        }

        if (request.getEmailNotifications() != null) {
            user.setEmailNotifications(request.getEmailNotifications());
        }
        if (request.getJobAlerts() != null) {
            user.setJobAlerts(request.getJobAlerts());
        }
        if (request.getApplicationUpdates() != null) {
            user.setApplicationUpdates(request.getApplicationUpdates());
        }
        if (request.getWeeklyNewsletter() != null) {
            user.setWeeklyNewsletter(request.getWeeklyNewsletter());
        }
        if (request.getProfileVisibility() != null && !request.getProfileVisibility().isBlank()) {
            user.setProfileVisibility(request.getProfileVisibility().trim());
        }
        if (request.getShowEmail() != null) {
            user.setShowEmail(request.getShowEmail());
        }
        if (request.getShowPhone() != null) {
            user.setShowPhone(request.getShowPhone());
        }
        if (request.getLanguage() != null && !request.getLanguage().isBlank()) {
            user.setLanguage(request.getLanguage().trim());
        }
        if (request.getTimezone() != null && !request.getTimezone().isBlank()) {
            user.setTimezone(request.getTimezone().trim());
        }

        return mapToSettings(userRepository.save(user));
    }

    public UserProfileResponse mapToProfile(User user) {
        return UserProfileResponse.builder()
                .userId(user.getId())
                .name(user.getName())
                .fullName(user.getName())
                .email(user.getEmail())
                .role(user.getRole().name())
                .resumeUrl(user.getResumeUrl())
                .phone(user.getPhone())
                .location(user.getLocation())
                .bio(user.getBio())
                .skills(user.getSkills())
                .experience(user.getExperience())
                .education(user.getEducation())
                .build();
    }

    private UserSettingsResponse mapToSettings(User user) {
        return UserSettingsResponse.builder()
                .email(user.getEmail())
                .emailNotifications(boolOrDefault(user.getEmailNotifications(), true))
                .jobAlerts(boolOrDefault(user.getJobAlerts(), true))
                .applicationUpdates(boolOrDefault(user.getApplicationUpdates(), true))
                .weeklyNewsletter(boolOrDefault(user.getWeeklyNewsletter(), false))
                .profileVisibility(user.getProfileVisibility() != null ? user.getProfileVisibility() : "public")
                .showEmail(boolOrDefault(user.getShowEmail(), false))
                .showPhone(boolOrDefault(user.getShowPhone(), false))
                .language(user.getLanguage() != null ? user.getLanguage() : "en")
                .timezone(user.getTimezone() != null ? user.getTimezone() : "Asia/Kolkata")
                .build();
    }

    private boolean boolOrDefault(Boolean value, boolean defaultValue) {
        return value == null ? defaultValue : value;
    }
}

File: jobGenieAi-frontend/src/services/userService.js
import apiClient from './apiClient';

const API_BASE = 'http://localhost:8080';

function normalizeProfile(data) {
  if (!data) return data;
  return {
    ...data,
    fullName: data.fullName || data.name || '',
    resumeUrl: data.resumeUrl
      ? data.resumeUrl.startsWith('http')
        ? data.resumeUrl
        : `${API_BASE}${data.resumeUrl}`
      : '',
  };
}

const userService = {
  async getCurrentProfile() {
    try {
      const response = await apiClient.get('/users/me');
      if (!response.ok) {
        throw new Error('Failed to fetch profile');
      }
      const profile = await response.json();
      return normalizeProfile(profile);
    } catch (error) {
      console.error('Error fetching profile:', error);
      throw error;
    }
  },

  async updateProfile(profile) {
    try {
      const body = {
        fullName: profile.fullName,
        phone: profile.phone,
        location: profile.location,
        bio: profile.bio,
        skills: profile.skills,
        experience: profile.experience,
        education: profile.education,
      };

      const response = await apiClient.put('/users/me', body);
      if (!response.ok) {
        const err = await response.json().catch(() => ({}));
        throw new Error(err.message || 'Failed to update profile');
      }
      const data = await response.json();
      return normalizeProfile(data);
    } catch (error) {
      console.error('Error updating profile:', error);
      throw error;
    }
  },

  async getSettings() {
    try {
      const response = await apiClient.get('/users/me/settings');
      if (!response.ok) {
        throw new Error('Failed to fetch settings');
      }
      return await response.json();
    } catch (error) {
      console.error('Error fetching settings:', error);
      throw error;
    }
  },

  async updateSettings(settings) {
    try {
      const body = {
        email: settings.email,
        emailNotifications: settings.emailNotifications,
        jobAlerts: settings.jobAlerts,
        applicationUpdates: settings.applicationUpdates,
        weeklyNewsletter: settings.weeklyNewsletter,
        profileVisibility: settings.profileVisibility,
        showEmail: settings.showEmail,
        showPhone: settings.showPhone,
        language: settings.language,
        timezone: settings.timezone,
      };

      if (settings.newPassword) {
        body.currentPassword = settings.currentPassword;
        body.newPassword = settings.newPassword;
        body.confirmPassword = settings.confirmPassword;
      }

      const response = await apiClient.put('/users/me/settings', body);
      if (!response.ok) {
        const err = await response.json().catch(() => ({}));
        throw new Error(err.message || 'Failed to update settings');
      }
      return await response.json();
    } catch (error) {
      console.error('Error updating settings:', error);
      throw error;
    }
  },

  async uploadResume(file) {
    try {
      const formData = new FormData();
      formData.append('resume', file);

      const response = await apiClient.post('/users/me/resume', formData, {
        headers: {},
      });

      if (!response.ok) {
        throw new Error('Failed to upload resume');
      }

      const data = await response.json();
      return normalizeProfile(data);
    } catch (error) {
      console.error('Error uploading resume:', error);
      throw error;
    }
  },
};

export default userService;
